pkg/config/config.go

```diff
@@ -0,0 +1,208 @@
+package config
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"math/rand"
+	"os"
+	"path/filepath"
+
+	"github.com/weaveworks/weave-gitops/pkg/logger"
+)
+
+const (
+	ConfigFileName       = "weave-gitops-config.json"
+	WrongConfigFormatMsg = `Your CLI configuration for Weave GitOps should represent a JSON object format:
+{
+	"analytics": true,
+	"userId": ""
+}
+Please set you configuration with: gitops set config`
+)
+
+var letters = []byte("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789+/")
+
+type GitopsCLIConfig struct {
+	Analytics bool   `json:"analytics"`
+	UserID    string `json:"userId"`
+}
+
+func (config *GitopsCLIConfig) String() (string, error) {
+	data, err := json.MarshalIndent(&config, "", "  ")
+	if err != nil {
+		return "", fmt.Errorf("error encoding config: %w", err)
+	}
+
+	return string(data), nil
+}
+
+// GetConfig reads the CLI configuration for Weave GitOps from the config file
+func GetConfig(log logger.Logger, shouldCreate bool) (*GitopsCLIConfig, error) {
+	log.Actionf("Reading GitOps CLI config ...")
+
+	configPath, err := getConfigPath(ConfigFileName)
+	if err != nil {
+		return nil, err
+	}
+
+	configFile, err := openConfigFile(configPath, shouldCreate)
+	if err != nil {
+		log.Failuref("Error opening config file at path: %s", configPath)
+		log.Warningf(WrongConfigFormatMsg)
+
+		return nil, err
+	}
+
+	defer configFile.Close()
+
+	config := &GitopsCLIConfig{}
+
+	data, err := readData(configFile)
+	if err != nil && !shouldCreate {
+		log.Failuref("Error reading config data from file at path: %s", configPath)
+		log.Warningf(WrongConfigFormatMsg)
+
+		return nil, err
+	}
+
+	if len(data) == 0 {
+		if !shouldCreate {
+			log.Warningf(WrongConfigFormatMsg)
+			return nil, fmt.Errorf("empty config file detected at path: %s", configPath)
+		}
+	} else if err = parseConfig(data, config); err != nil {
+		log.Failuref("Error reading GitOps CLI config from file")
+
+		if shouldCreate {
+			// just replace invalid config with default config
+			log.Actionf("Replacing invalid config ...")
+		} else {
+			log.Warningf(WrongConfigFormatMsg)
+			return nil, err
+		}
+	}
+
+	return config, nil
+}
+
+// SaveConfig saves the CLI configuration for Weave GitOps to the config file
+func SaveConfig(log logger.Logger, config *GitopsCLIConfig) error {
+	log.Actionf("Saving GitOps CLI config ...")
+
+	configPath, err := getConfigPath(ConfigFileName)
+	if err != nil {
+		return err
+	}
+
+	data, err := json.MarshalIndent(&config, "", "  ")
+	if err != nil {
+		return fmt.Errorf("error encoding config: %w", err)
+	}
+
+	configFile, err := openConfigFile(configPath, true)
+	if err != nil {
+		return fmt.Errorf("error opening config file: %w", err)
+	}
+
+	defer configFile.Close()
+
+	err = configFile.Truncate(0)
```

Comment:
:warning: You know you just opened the file, so you don't need to truncate and seek - you would need to explicitly open it in append mode for "open(); write()" to not do what you are doing here.

:bulb: It's common for data encoding and decoding APIs to have a way to use a IO reader or writer (which is to say, an open file) to stream data to/from - in the json module, there's an `Encoder` and a `Decoder` that you could use directly with the file descriptor. This might have been important if the data could be "big" (since it saves copying the data), but for this it doesn't really matter.

Reply:
Thanks, good to know!

Reply:
Removed unneeded calls to `Truncate` and `Seek`.